src/memo/memo.mli

```diff
@@ -29,19 +44,59 @@ end
     are forgotten, pending computations are cancelled. *)
 val reset : unit -> unit
 
-module type S = Memo_intf.S with type stack_frame := Stack_frame.t
 module type Input = Memo_intf.Input
-module type Output = Memo_intf.Output
-module type Decoder = Memo_intf.Decoder
+module type Data = Memo_intf.Data
 
-module Make(Input : Input)(Decoder : Decoder with type t := Input.t)
-  : S with type input := Input.t
+module Function : sig
+  type ('a, 'b, 'f) t =
+    | Sync : ('a -> 'b) -> ('a, 'b, ('a -> 'b)) t
+    | Async : ('a -> 'b Fiber.t) -> ('a, 'b, ('a -> 'b Fiber.t)) t
+end
+
+module Function_type : sig
+  type ('a, 'b, 'f) t =
+    | Sync : ('a, 'b, ('a -> 'b)) t
+    | Async : ('a, 'b, ('a -> 'b Fiber.t)) t
+end
+
+module Output : sig
+  type 'o t =
+    | Simple of (module Memo_intf.Sexpable with type t = 'o)
```

Comment:
I know this PR doesn't introduce cutoff, but it wouldn't hurt documenting what it is here.

Reply:
Added the following doc

````
(**
   When we recompute the function and find that its output is the same as what we
   computed before, we can sometimes skip recomputing the values that depend on it.

   [Allow_cutoff] specifies how to compare the output values for that purpose.

   Note that currently Dune wipes all memoization caches on every run, so
   cutoff is not effective.

   (* CR-someday aalekseyev:
   Not sure why the type is so complicated.
   Couldn't we just pass an equality function as an extra optional argument to [create]?
   We don't use hash anywhere, I think. *)
*)
````

Should I "do" the CR right now? (yeah, I should have thought about this before)

Reply:
Either way works. I think it's fine to leave things as is and simplify in a subsequent PR.

Reply:
Let's cc @diml here regarding why `Allow_cutoff` is so complex. I don't see where the hashing is used.

Reply:
I removed `hash` and removed this CR.
The reason I'm keeping the first-class module instead of a function is that this makes the common case of `Allow_cutoff`  shorter: you don't need to repeat yourself by writing `~output:(module Foo.Bar.Baz) ~allow_cutoff:(Foo.Bar.Baz.equal)`.